File: data-structures/stacks/c-sharp/stack.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace IsaacCodeSamples
{
    class Stacks
    {
        public const int MaxSize = 4;

        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            Console.WriteLine("### Stack ###");

            // Initialise the stack and pointers
            string[] stack = new string[MaxSize];
            int top = -1;

            // Insert test data into the stack
            top = Push(stack, top, "Julie");
            OutputStack(stack, top);
            
            top = Push(stack, top, "Rey");
            OutputStack(stack, top);
            
            top = Push(stack, top, "Habib");
            OutputStack(stack, top);
            
            top = Push(stack, top, "Sabrina");
            OutputStack(stack, top);

            // Trying to push - should say stack is full
            top = Push(stack, top, "Mina");
            OutputStack(stack, top);

            // Peek the top item of the stack
            string peekedItem = Peek(stack, top);
            Console.Write($"\nPeeked top of the stack {peekedItem}");
            OutputStack(stack, top);

            // Pop all of the items
            foreach (string item in stack) {
                // Tuple is used to store multiple return values
                Tuple<string, int> poppedValues; // item, top

                poppedValues = Pop(stack, top);
                top = poppedValues.Item2;
                Console.Write($"\nPopped {poppedValues.Item1}");
                OutputStack(stack, top);
            }

            // Try to pop - should say stack is empty
            Tuple<string, int> finalPoppedValues; // item, top
            finalPoppedValues = Pop(stack, top);
            top = finalPoppedValues.Item2;
        }

        // Check if the stack is empty
        public static bool IsEmpty(int top)
        {
            if (top == -1)
                return true;
            else
                return false;
        }

        // Check if the stack is full
        public static bool IsFull(int top)
        {
            if (top == MaxSize - 1)
                return true;
            else
                return false; 
        }

        // Push data onto the top of the stack
        public static int Push(string[] stack, int top, string data)
        {
            if (IsFull(top) == true)
                Console.WriteLine($"\nStack is full - {data} not added");
            else {
                top = top + 1;
                stack[top] = data;
            }
            return top;
        }

        // Return a copy of the item from the top of stack without removing it
        public static string Peek(string[] stack, int top)
        {
            string peekedItem;

            if (IsEmpty(top) == true) {
                Console.WriteLine("\nStack is empty - nothing to peek");
                peekedItem = "";
            }
            else {
                peekedItem = stack[top];
            }

            return peekedItem;
        }

        // Return the item from the top of stack and remove it from the stack
        public static Tuple<string, int> Pop(string[] stack, int top)
        {
            string poppedItem;

            if (IsEmpty(top) == true) {
                Console.WriteLine("\nStack is empty - nothing to pop");
                poppedItem = "";
            }
            else {
                poppedItem = stack[top];
                stack[top] = ""; // Testing
                top = top - 1;
            }

            return Tuple.Create(poppedItem, top);
        }

        // Output the state of the stack
        public static void OutputStack(string[] stack, int top)
        {
            // Testing
            Console.WriteLine("\n------ State of the stack (first item is the top) ------");

            for (int i = MaxSize - 1; i >= 0; i--) {
                string item = stack[i];
                if (string.IsNullOrEmpty(item)) {
                    item = "null";
                }
                Console.WriteLine($"{item}");
            }

            Console.WriteLine($"Top pointer: {top}");
        }

    }
}

File: data-structures/stacks/python/stack.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4


MAX_SIZE = 4 # Max size of the stack


def is_empty(top):
    """Check if the stack is empty"""
    
    if top == -1:
        return True
    else:
        return False  


def is_full(top):
    """Check if the stack is full"""
    
    if top == MAX_SIZE - 1:
        return True
    else:
        return False


def push(stack, top, data):
    """Push data onto the top of the stack"""
    
    if is_full(top) == True:
        print(f"Stack is full - {data} not added")
    else:
        top = top + 1
        stack[top] = data
    return top      


def peek(stack, top):
    """Return a copy of the item from the top of stack without removing it"""
    
    if is_empty(top):
        print("Stack is empty - nothing to peek")
        peeked_item = None
    else:
        peeked_item = stack[top]
    return peeked_item


def pop(stack, top):
    """Return the item from the top of stack and remove it from the stack"""
    
    if is_empty(top) == True:
        print("\nStack is empty - nothing to pop")
        popped_item = None
    else:
        popped_item = stack[top]
        stack[top] = None # Testing
        top = top - 1
        
    return popped_item, top


def output_stack(stack, top):
    """Output the state of the stack"""
    
    print("\n------ State of the stack (first item is the top) ------")
    for i in range(MAX_SIZE - 1, -1, -1):
        print(stack[i])
    print(f"Top pointer: {top}")


def main():
    """Create a stack and push, pop and peek items"""
    
    print("### Stack ###")

    # Initialise the stack and pointers
    stack = []
    top = -1

    # Simulate a fixed size array
    for i in range(MAX_SIZE):
        stack.append(None)
    output_stack(stack, top)

    # Insert test data onto the stack
    top = push(stack, top, "Julie")
    output_stack(stack, top)
    
    top = push(stack, top, "Rey")
    output_stack(stack, top)
    
    top = push(stack, top, "Habib")
    output_stack(stack, top)
    
    top = push(stack, top, "Sabrina")
    output_stack(stack, top)

    # Trying to push - should say stack is full
    top = push(stack, top, "Mina")
    output_stack(stack, top)

    # Peek the top item of the stack
    peeked_item = peek(stack, top)
    print(f"\nPeeked top of the stack: {peeked_item}", end="")
    output_stack(stack, top)

    # Pop all of the items
    for item in stack:
        item, top = pop(stack, top)
        print(f"\nPopped {item}", end="")
        output_stack(stack, top)

    # Try to pop - should say stack is empty
    item, top = pop(stack, top)
    print(f"Popped {item}", end="")
    

# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == '__main__':
    main()
